Check the permission of the underlying storage

rullzer opened the pull request:
Else shares might expose more permissions than the storage actually
providers.

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Check the permission of the underlying storage

Else shares might expose more permissions than the storage actually
providers.

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
Ok so the reason it does boom is because of https://github.com/nextcloud/server/blob/0eebff152a177dd59ed8773df26f1679f8a88e90/lib/private/Files/Cache/Wrapper/CacheWrapper.php#L96

Which is the function called from the shared cache. This gives back the actual path instead of the jailed one. Making the getPermissions do :boom:.

@icewind1991 do we have to fixed the shared cache for this?

rullzer committed:
Fix shared cache

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
@icewind1991 please have a look if https://github.com/nextcloud/server/pull/11041/commits/74971a22730fe8f52295e4f3e1795df49386e8c4 makes sense.

It does to me but you know more about the internal dark magic of the storages.

icewind1991 commented on apps/files_sharing/lib/Cache.php:
```diff
@@ -169,4 +169,9 @@ private function getOwnerDisplayName() {
 		// Not a valid action for Shared Cache
 	}
+
+	public function getFolderContentsById($fileId) {
```
this shouldn't be needed as it's the same as the base method

rullzer replied:
Hehe late night hacking ;)

rullzer commented on apps/files_sharing/lib/Cache.php:
```diff
@@ -142,7 +142,7 @@ protected function formatCacheEntry($entry, $path = null) {
 			$entry['path'] = $path;
 		}
-		$sharePermissions = $this->storage->getPermissions($path);
+		$sharePermissions = $this->storage->getPermissions($entry['path']);
```
So this one actually is correct right @icewind1991

blizzz replied:
this caused https://github.com/nextcloud-gmbh/sysadmin/issues/156 on a valid file (mounted public share from the same instance). "Fixed" with #11383, which originally was to catch an edge case. Eventually, it seems this is not acting 100% as it should.

nickvergessen reviewed (approved):
Still works

rullzer commented:
It seems this makes other stuff do :boom:. (yay for part files).
I'll try to come up with a fix

rullzer committed:
Properly check share permissions

isCreatable only works on folders
isUpdatable if the file is not there but it is a part file also has to
be checked on the folder

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Update test now that we check permissions properly

Now that we actually check thepermissions properly we have to update the
tests.

* We checked an invalid path
* We checked from wrong permissions (files never have CREATE permissions
for example)

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
Ok all good now I think...

@nickvergessen @icewind1991 please have yet another look.

nickvergessen commented:
@icewind1991 your +1 still counts?

icewind1991 commented:
:+1:
